### PR TITLE
Tree widget

### PR DESCRIPTION
**Pull request by namreg**
Hello, there.
I've created a tree widget for my project. I believe this can be useful for anyone else :)
Let me know, please, what you think about that.

Thanks for the termui project!

### CODE REVIEW

**Comment by srlehn** [comment]
screenshot of the example:
Auswahl_2019 04 19_12:07:52

**Comment by cjbassi** [comment]
Thanks for the pr! I'd like to merge this but I'll need some time to look over it and possibly suggest any changes.

File: theme.go [hunk 1]
```diff
@@ -66,6 +67,12 @@ type ListTheme struct {
 	Text Style
 }
 
+type TreeTheme struct {
+	Text     Style
+	Collapse rune
+	Expand   rune
```
Comment (cjbassi) [inline theme.go]: Rename `Collapse` to `Collapsed` and `Expand` to `Expanded`.

File: widgets/tree.go [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+	return &Tree{
+		Block:            *NewBlock(),
+		TextStyle:        Theme.Tree.Text,
+		SelectedRowStyle: Theme.Tree.Text,
+		WrapText:         true,
+	}
+}
+
+func (self *Tree) prepareNodes() {
+	self.rows = self.rows[:0]
```
Comment (cjbassi) [inline widgets/tree.go]: Is this emptying the slice? A comment explaining this would be good.

File: widgets/tree.go [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+type Tree struct {
+	Block
+	Nodes            []*TreeNode
+	TextStyle        Style
+	SelectedRowStyle Style
+	WrapText         bool
+	SelectedRow      int
+	topRow           int
+
+	//rows is flatten nodes for rendering.
```
Comment (cjbassi) [inline widgets/tree.go]: Fix space before comment text.

File: widgets/tree.go [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+
+const treeIndent = "  "
+
+// TreeNode is a tree node.
+type TreeNode struct {
+	Value    fmt.Stringer
+	Expanded bool
+	Nodes    []*TreeNode
+
+	//level stores the node level in the tree.
```
Comment (cjbassi) [inline widgets/tree.go]: Fix space before comment text.

File: widgets/tree.go [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+
+// Tree is a tree widget.
+type Tree struct {
+	Block
+	Nodes            []*TreeNode
+	TextStyle        Style
+	SelectedRowStyle Style
+	WrapText         bool
+	SelectedRow      int
+	topRow           int
```
Comment (cjbassi) [inline widgets/tree.go]: Move this field down a line next to the other private fields.

**Comment by namreg** [comment]
@cjbassi 
> 1. How should the cursor be positioned after collapsing a node? Currently it seems like SelectedRow stays the same, but sometimes that leaves no currently visible node selected. Although that might be fine. Just wanted to get some thoughts.

I did it intentionally. IMHO, cursor position relate to the line in block area, not to the content is rendered to this area. May be I wrong.

> 2. Expand and the other functions that operate on self.rows would potentially operate on an incorrect node if a node is added and Expand is called before the tree is rerendered, since self.rows is only updated in Draw. What if we made self.Nodes private and added a SetNodes method that calls prepareNodes? That would also eliminate calling prepareNodes every draw too.

It like the idea to add public method `SetNodes` and make property `rows` private. But we can't rid of the `prepareNodes` calling before render after that because `self.rows` contains only visible rows (in our case `expanded`). Hence, I have moved calling of the `prepareNodes` to every place where `self.rows` is changing. Please, let me know what you think about that.

**Comment by namreg** [comment]
gentle ping @cjbassi

**Comment by cjbassi** [comment]
Apologies for the delay, but I haven't forgotten about this! Currently finishing up the school quarter and I should be able to get to this soon.

**Comment by namreg** [comment]
No problem) Good luck in the school!

**Comment by cjbassi** [comment]
Sorry for the delay again! Been out of commission a bit due to an injury. I've looked it over again and it LGTM. Changing the widget so that nodes are set with `SetNodes` looks good. I had to rebase to fix some conflicts but I just merged this and now I'm going to release 3.1. Thanks!

**Comment by cjbassi** [comment]
I accidentally merged before force pushing the PR which is why it says the PR was closed just btw.

**Pull request closed without merging**